File: domain-persistence-modeling/java-spring-boot/src/main/java/x/scratch/parent/ParentRecord.java
package x.scratch.parent;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Table;
import x.scratch.UpsertableRecord;

import java.util.Set;
import java.util.TreeSet;

import static lombok.AccessLevel.PACKAGE;

@AllArgsConstructor(access = PACKAGE)
@Builder(toBuilder = true)
@Data
@Table("parent")
public class ParentRecord
        implements ParentSimpleDetails,
        UpsertableRecord<ParentRecord> {
    @Id
    private Long id;
    private @NonNull String naturalId;
    private String value;
    @Builder.Default
    private @NonNull Set<String> sideValues = new TreeSet<>();
    private int version;

    static ParentRecord createRecordFor(final String naturalId) {
        return new ParentRecord(null, naturalId, null, new TreeSet<>(), 0);
    }

    @Override
    public ParentRecord upsertedWith(final ParentRecord upserted) {
        id = upserted.id;
        naturalId = upserted.naturalId;
        value = upserted.value;
        version = upserted.version;
        return this;
    }
}

File: domain-persistence-modeling/kotlin-micronaut/src/main/kotlin/x/domainpersistencemodeling/parent/ParentRecord.kt
package x.domainpersistencemodeling.parent

import io.micronaut.core.annotation.Introspected
import x.domainpersistencemodeling.KnownState.ENABLED
import x.domainpersistencemodeling.UpsertableRecord
import java.util.TreeSet
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.Table

@Entity
@Introspected
@Table(name = "parent")
data class ParentRecord(
    @Id var id: Long?,
    override var naturalId: String,
    override var otherNaturalId: String?,
    override var state: String,
    override var value: String?,
    override var sideValues: MutableSet<String>,
    override var version: Int
) : MutableParentSimpleDetails,
    UpsertableRecord<ParentRecord> {
    internal constructor(naturalId: String)
            : this(
        null, naturalId, null, ENABLED.name, null,
        mutableSetOf(),
        0
    )

    override fun upsertedWith(upserted: ParentRecord): ParentRecord {
        id = upserted.id
        naturalId = upserted.naturalId
        otherNaturalId = upserted.otherNaturalId
        state = upserted.state
        value = upserted.value
        sideValues = TreeSet(upserted.sideValues)
        version = upserted.version
        return this
    }
}
